File: code/R/chapter03_cleaning.R
# Chapter 3 reproducible data-preparation pipeline
# Run from the project root.

library(readr)
library(dplyr)
library(stringr)

dat <- read_csv(
  "data/chapter03_raw_student_data.csv",
  col_types = cols(.default = col_character()),
  na = character()
)
raw_rows <- nrow(dat)

dat <- dat |>
  mutate(
    across(everything(), str_trim),
    student_id = str_to_upper(student_id),
    record_version = suppressWarnings(as.numeric(record_version))
  )

unique_ids <- n_distinct(dat$student_id)
extra_duplicate_rows <- raw_rows - unique_ids

dat <- dat |>
  arrange(student_id, record_version) |>
  group_by(student_id) |>
  slice_tail(n = 1) |>
  ungroup()

normalise_key <- function(x) {
  x |>
    str_trim() |>
    str_to_upper() |>
    str_replace_all("[-_]", " ")
}

date_text <- str_replace_all(dat$entry_date, "/", "-")
dat$entry_date_clean <- suppressWarnings(as.Date(date_text))

dat$age_clean <- suppressWarnings(as.numeric(dat$age))
dat$age_clean[
  is.na(dat$age_clean) |
  dat$age_clean < 18 |
  dat$age_clean > 75
] <- NA_real_

dat$study_hours_clean <- suppressWarnings(as.numeric(dat$study_hours))
dat$study_hours_clean[
  is.na(dat$study_hours_clean) |
  dat$study_hours_clean < 0 |
  dat$study_hours_clean > 112
] <- NA_real_

dat$exam_score_clean <- suppressWarnings(as.numeric(dat$exam_score))
dat$exam_score_clean[
  is.na(dat$exam_score_clean) |
  dat$exam_score_clean < 0 |
  dat$exam_score_clean > 100
] <- NA_real_

gender_key <- normalise_key(dat$gender)
dat$gender_clean <- case_when(
  gender_key %in% c("F", "FEMALE", "WOMAN") ~ "Female",
  gender_key %in% c("M", "MALE", "MAN") ~ "Male",
  gender_key %in% c("NB", "NON BINARY", "NONBINARY") ~ "Nonbinary",
  gender_key %in% c("PNS", "PREFER NOT TO SAY") ~ "Prefer not to say",
  TRUE ~ NA_character_
)

format_key <- normalise_key(dat$instructional_format)
dat$instructional_format_clean <- case_when(
  format_key %in% c("F2F", "FACE TO FACE") ~ "Face-to-face",
  format_key %in% c("ONLINE", "REMOTE") ~ "Online",
  format_key %in% c("HYB", "HYBRID") ~ "Hybrid",
  TRUE ~ NA_character_
)

completion_key <- normalise_key(dat$completion_status)
dat$completion_status_clean <- case_when(
  completion_key %in% c("Y", "1", "COMPLETED", "COMPLETE") ~ "Completed",
  completion_key %in% c("N", "0", "DID NOT COMPLETE", "NOT COMPLETED") ~
    "Did not complete",
  TRUE ~ NA_character_
)

consent_key <- normalise_key(dat$consent)
dat$consent_clean <- case_when(
  consent_key %in% c("Y", "YES", "1") ~ "Yes",
  consent_key %in% c("N", "NO", "0") ~ "No",
  TRUE ~ NA_character_
)

for (j in 1:5) {
  source <- paste0("item", j)
  target <- paste0(source, "_clean")
  values <- suppressWarnings(as.numeric(dat[[source]]))
  values[is.na(values) | values < 1 | values > 5] <- NA_real_
  dat[[target]] <- values
}

dat <- dat |>
  mutate(
    item3_reverse = 6 - item3_clean,
    item5_reverse = 6 - item5_clean
  )

scored <- dat |>
  select(item1_clean, item2_clean, item3_reverse,
         item4_clean, item5_reverse)

dat$scale_items_observed <- rowSums(!is.na(scored))
dat$self_efficacy_score <- rowMeans(scored, na.rm = TRUE)
dat$self_efficacy_score[
  dat$scale_items_observed < 4
] <- NA_real_

dat$age_group <- cut(
  dat$age_clean,
  breaks = c(17, 24, 34, 44, 75),
  labels = c("18-24", "25-34", "35-44", "45-75")
)

analysis <- dat |>
  filter(consent_clean == "Yes") |>
  transmute(
    student_id,
    record_version,
    entry_date = entry_date_clean,
    age = age_clean,
    age_group,
    gender = gender_clean,
    instructional_format = instructional_format_clean,
    study_hours = study_hours_clean,
    item1 = item1_clean,
    item2 = item2_clean,
    item3 = item3_clean,
    item3_reverse,
    item4 = item4_clean,
    item5 = item5_clean,
    item5_reverse,
    scale_items_observed,
    self_efficacy_score,
    exam_score = exam_score_clean,
    completion_status = completion_status_clean,
    consent = consent_clean
  )

write_csv(
  analysis,
  "data/chapter03_clean_student_data_R.csv"
)

audit <- tibble(
  audit_item = c(
    "Raw rows imported",
    "Unique normalised student identifiers",
    "Extra rows caused by repeated identifiers",
    "Rows retained after keeping highest record version",
    "Rows excluded because consent was not Yes",
    "Rows in final analysis data"
  ),
  count = c(
    raw_rows,
    unique_ids,
    extra_duplicate_rows,
    nrow(dat),
    sum(dat$consent_clean != "Yes" | is.na(dat$consent_clean)),
    nrow(analysis)
  )
)

write_csv(audit, "support/chapter03_audit_R.csv")
print(audit)

File: code/Python/chapter03_cleaning.py
"""Chapter 3 reproducible data-preparation pipeline.

Run from the project root:
    python code/Python/chapter03_cleaning.py
"""

from pathlib import Path
import numpy as np
import pandas as pd

RAW = Path("data/chapter03_raw_student_data.csv")
CLEAN = Path("data/chapter03_clean_student_data_python.csv")
AUDIT = Path("support/chapter03_audit_python.csv")

def canonical_key(value):
    if pd.isna(value):
        return pd.NA
    return (
        str(value).strip().upper()
        .replace("-", " ")
        .replace("_", " ")
    )

dat = pd.read_csv(RAW, dtype="string", keep_default_na=False)
raw_rows = len(dat)

for column in dat.columns:
    dat[column] = dat[column].str.strip()

dat["student_id"] = dat["student_id"].str.upper()
dat["record_version"] = pd.to_numeric(
    dat["record_version"], errors="coerce"
)

unique_ids = dat["student_id"].nunique(dropna=True)
extra_duplicate_rows = raw_rows - unique_ids

dat = (
    dat.sort_values(["student_id", "record_version"])
       .drop_duplicates("student_id", keep="last")
       .reset_index(drop=True)
)

date_text = dat["entry_date"].str.replace("/", "-", regex=False)
dat["entry_date_clean"] = pd.to_datetime(
    date_text, errors="coerce"
).dt.date

for source, target, lower, upper in [
    ("age", "age_clean", 18, 75),
    ("study_hours", "study_hours_clean", 0, 112),
    ("exam_score", "exam_score_clean", 0, 100),
]:
    dat[target] = pd.to_numeric(dat[source], errors="coerce")
    dat.loc[~dat[target].between(lower, upper), target] = np.nan

gender_map = {
    "F": "Female", "FEMALE": "Female", "WOMAN": "Female",
    "M": "Male", "MALE": "Male", "MAN": "Male",
    "NB": "Nonbinary", "NON BINARY": "Nonbinary",
    "NONBINARY": "Nonbinary",
    "PNS": "Prefer not to say",
    "PREFER NOT TO SAY": "Prefer not to say",
}
format_map = {
    "F2F": "Face-to-face",
    "FACE TO FACE": "Face-to-face",
    "ONLINE": "Online", "REMOTE": "Online",
    "HYB": "Hybrid", "HYBRID": "Hybrid",
}
completion_map = {
    "Y": "Completed", "1": "Completed",
    "COMPLETED": "Completed", "COMPLETE": "Completed",
    "N": "Did not complete", "0": "Did not complete",
    "DID NOT COMPLETE": "Did not complete",
    "NOT COMPLETED": "Did not complete",
}
consent_map = {
    "Y": "Yes", "YES": "Yes", "1": "Yes",
    "N": "No", "NO": "No", "0": "No",
}

dat["gender_clean"] = dat["gender"].map(
    lambda x: gender_map.get(canonical_key(x), pd.NA)
)
dat["instructional_format_clean"] = dat[
    "instructional_format"
].map(lambda x: format_map.get(canonical_key(x), pd.NA))
dat["completion_status_clean"] = dat[
    "completion_status"
].map(lambda x: completion_map.get(canonical_key(x), pd.NA))
dat["consent_clean"] = dat["consent"].map(
    lambda x: consent_map.get(canonical_key(x), pd.NA)
)

for j in range(1, 6):
    source = f"item{j}"
    target = f"{source}_clean"
    dat[target] = pd.to_numeric(dat[source], errors="coerce")
    dat.loc[~dat[target].between(1, 5), target] = np.nan

dat["item3_reverse"] = 6 - dat["item3_clean"]
dat["item5_reverse"] = 6 - dat["item5_clean"]

scored_items = [
    "item1_clean", "item2_clean", "item3_reverse",
    "item4_clean", "item5_reverse",
]
dat["scale_items_observed"] = dat[scored_items].notna().sum(axis=1)
dat["self_efficacy_score"] = dat[scored_items].mean(axis=1)
dat.loc[
    dat["scale_items_observed"] < 4,
    "self_efficacy_score",
] = np.nan

dat["age_group"] = pd.cut(
    dat["age_clean"],
    bins=[17, 24, 34, 44, 75],
    labels=["18-24", "25-34", "35-44", "45-75"],
)

analysis = dat.loc[dat["consent_clean"] == "Yes"].copy()

output = analysis[[
    "student_id", "record_version", "entry_date_clean",
    "age_clean", "age_group", "gender_clean",
    "instructional_format_clean", "study_hours_clean",
    "item1_clean", "item2_clean", "item3_clean",
    "item3_reverse", "item4_clean", "item5_clean",
    "item5_reverse", "scale_items_observed",
    "self_efficacy_score", "exam_score_clean",
    "completion_status_clean", "consent_clean",
]].rename(columns={
    "entry_date_clean": "entry_date",
    "age_clean": "age",
    "gender_clean": "gender",
    "instructional_format_clean": "instructional_format",
    "study_hours_clean": "study_hours",
    "item1_clean": "item1",
    "item2_clean": "item2",
    "item3_clean": "item3",
    "item4_clean": "item4",
    "item5_clean": "item5",
    "exam_score_clean": "exam_score",
    "completion_status_clean": "completion_status",
    "consent_clean": "consent",
})

output.to_csv(CLEAN, index=False)

audit = pd.DataFrame({
    "audit_item": [
        "Raw rows imported",
        "Unique normalised student identifiers",
        "Extra rows caused by repeated identifiers",
        "Rows retained after keeping highest record version",
        "Rows excluded because consent was not Yes",
        "Rows in final analysis data",
    ],
    "count": [
        raw_rows,
        unique_ids,
        extra_duplicate_rows,
        len(dat),
        int((dat["consent_clean"] != "Yes").sum()),
        len(output),
    ],
})
audit.to_csv(AUDIT, index=False)
print(audit.to_string(index=False))